File: leecode/leetcode-cn/1610.可见点的最大数目.java
// 方法一：排序 + 二分查找
class Solution {
    public int visiblePoints(List<List<Integer>> points, int angle, List<Integer> location) {
        int sameCnt = 0;
        int maxCnt = 0;

        List<Double> polarRadians = new ArrayList<>();
        for (List<Integer> point : points) {
            int y = point.get(1) - location.get(1);
            int x = point.get(0) - location.get(0);
            if (y == 0 && x == 0) {
                sameCnt++;
                continue;
            }
            // tip1: atan2值域范围[-pi,pi]
            polarRadians.add(Math.atan2(y, x)); 
        }

        // 排序
        Collections.sort(polarRadians);

        int m = polarRadians.size();
        // tip2: 环形处理，将原有数组加上2*pi放在后面，防止
        for (int i = 0; i < m; i++) {
            polarRadians.add(polarRadians.get(i) + 2 * Math.PI);
        }

        // tip3: 二分查找，找出 [di,di+angle] 内的点
        double toRadian = angle * Math.PI / 180;

        for (int i = 0; i < m; i++) {
            int right = binarySearch(polarRadians, polarRadians.get(i) +  toRadian, false);
            maxCnt = Math.max(maxCnt, right - i);
        }

        return sameCnt + maxCnt;
    }

    // 如果lower为false，与C++ upper_bound()函数相同；
    // 反之与C++ lower_bound()相同
    public int binarySearch(List<Double> nums, Double target, boolean lower) {
        int left = 0, right = nums.size() - 1;
        int ans = nums.size(); // 注意
        while(left <= right) {
          int mid = (left + right) / 2;
          if(nums.get(mid) > target || (lower && nums.get(mid) >= target)) {
            right = mid - 1;
            ans = mid;
          } else {
            left = mid + 1;
          }
        }
        return ans;
    }
}


// 方法二：排序 + 滑动窗口
// 该方法与上面的方法基本相同，唯一不同的是使用滑动窗口使得区间查找的时间优化为O(2n+2n)
// 时间复杂度：O(nlogn + 2n+2n)
class Solution {
    public int visiblePoints(List<List<Integer>> points, int angle, List<Integer> location) {
        int sameCnt = 0;
        int maxCnt = 0;

        List<Double> polarRadians = new ArrayList<>();
        for (List<Integer> point : points) {
            int y = point.get(1) - location.get(1);
            int x = point.get(0) - location.get(0);
            if (y == 0 && x == 0) {
                sameCnt++;
                continue;
            }
            // tip1: atan2值域范围[-pi,pi]
            polarRadians.add(Math.atan2(y, x)); 
        }

        // 排序
        Collections.sort(polarRadians);

        int m = polarRadians.size();
        // tip2: 环形处理，将原有数组加上2*pi放在后面，防止
        for (int i = 0; i < m; i++) {
            polarRadians.add(polarRadians.get(i) + 2 * Math.PI);
        }

        // tip3: 滑动窗口进行区间查找，找出 [di,di+angle] 内的点
        int right = 0;
        double toRadian = angle * Math.PI / 180;

        for (int i = 0; i < m; i++) {
            double curr = polarRadians.get(i);
            while (right < polarRadians.size() && polarRadians.get(right) - curr <= toRadian) {
                right++;
            }
            maxCnt = Math.max(maxCnt, right - i);
        }

        return sameCnt + maxCnt;
    }
}

File: leecode/leetcode-cn/1610.可见点的最大数目.cpp
/*
 * @lc app=leetcode.cn id=1610 lang=cpp
 *
 * [1610] 可见点的最大数目
 */

// @lc code=start
// 1. 全部化成极坐标
// 2. atan2(y, x) 求 y/x 的反正切，值域为 [-π, π]
// 3. 需要环形遍历：例如：{0->1, 1->2} 没有遍历完，还需要遍历 {2->0} 。
//    可以增加一个周期，以避免边界讨论。
// 方法一：滑动窗口
class Solution {
public:
    int visiblePoints(vector<vector<int>>& points, int angle, vector<int>& location) {
        vector<double> radians;
        // 如果点处于极坐标原点，需要单独处理（此时 atan2 的返回值也可能不确定）。
        int ans = 0, same = 0;
        for (auto& p : points) {
            int px = p[0] - location[0], py = p[1] - location[1];
            if (px == 0 && py == 0)
                ++same;
            else
                radians.push_back(atan2(py, px));
        }
        ranges::sort(radians);
        // 增加一个周期
        int m = radians.size();
        for (int i = 0; i < m; ++i) {
            radians.push_back(radians[i] + M_PI * 2);
        }
        double radian = angle * M_PI / 180;
        // 滑动窗口
        // 遍历每个左端点，右端点尽可能伸展，这样才能走完一个完整环
        for (int l = 0, r = 0; l < m; ++l) {
            while (r < radians.size() && radians[l] + radian >= radians[r])
                ++r;
            ans = max(ans, r - l);
        }
        return ans + same;
    }
};
// @lc code=end

// 方法二：二分查找，与滑动窗口本质相同，每次二分查找右边界 d + angle
namespace solution_2 {

class Solution {
public:
    int visiblePoints(vector<vector<int>>& points, int angle, vector<int>& location) {
        vector<double> radians;
        // 如果点处于极坐标原点，需要单独处理（此时 atan2 的返回值也可能不确定）。
        int ans = 0, same = 0;
        for (auto& p : points) {
            int px = p[0] - location[0], py = p[1] - location[1];
            if (px == 0 && py == 0)
                ++same;
            else
                radians.push_back(atan2(py, px));
        }
        ranges::sort(radians);
        // 增加一个周期
        int m = radians.size();
        for (int i = 0; i < m; ++i) {
            radians.push_back(radians[i] + M_PI * 2);
        }
        double radian = angle * M_PI / 180;
        // 遍历左端点，二分查找右端点
        for (int l = 0; l < m; ++l) {
            auto it = upper_bound(radians.begin(), radians.end(), radians[l] + radian);
            int r = it - radians.begin();
            ans = max(ans, r - l);
        }
        return ans + same;
    }
};

}  // namespace solution_2
